Actual Report for Reference Setup, Performance, Scalability, and Sizing Guidelines (#7905)
longhorn/longhorn: examples/reference-setup-performance-scalability-and-sizing-guidelines/public-cloud/medium-node-spec.md

Reviewer: NIT: In many bulleted lists you use full sentences  and periods and in many you do not. It is better to be consistent where practical.

Author: Fixed. Thank you

# Reference Setup, Performance, Scalability, and Sizing Guidelines: Public Cloud - Medium Node Spec

## Table of contents
1. [Cluster Spec](#cluster-spec)
   1. [Node Spec](#node-spec)
   1. [Network Spec](#network-spec)
   1. [Disk Spec](#disk-spec)
   1. [Kubernetes Spec](#kubernetes-spec)
   1. [Longhorn Config](#longhorn-config)
   1. [Additional Components](#additional-components)

> 1. [Workload design](#workload-design)

Reviewer: NIT: This whole section seems to be generic for all cluster specs. Can we move it out of the medium-node-spec.md file?

Author: Since we only have 1 report at the moment. I think we can keep it here. Once we have multiple reports, we can refactor it later

   1. [Random Write IOPs - Rate Limited](#random-write-iops---rate-limited)
   1. [Sequential Write IOPs - Stress Tests](#sequential-write-iops---stress-tests)
   1. [Sequential Write IOPs - Rate Limited](#sequential-write-iops---rate-limited)
   1. [Random Write Bandwidth - Stress Tests](#random-write-bandwidth---stress-tests)
   1. [Random Write Bandwidth - Rate Limited](#random-write-bandwidth---rate-limited)
   1. [Sequential Write Bandwidth - Stress Tests](#sequential-write-bandwidth---stress-tests)
   1. [Sequential Write Bandwidth - Rate Limited](#sequential-write-bandwidth---rate-limited)
   1. [Random Write Latency - Stress Tests](#random-write-latency---stress-tests)
   1. [Sequential Write Latency - Stress Tests](#sequential-write-latency---stress-tests)
1. [Longhorn Control Plane Performance](#longhorn-control-plane-performance)
1. [Volume Maximum Actual Size](#volume-maximum-actual-size)

## Cluster Spec
### Node Spec
* EC2 instance type: ec2 m5zn.2xlarge
* 8vCPUs, 32GB RAM
* Root disk
  * Size 50GB 
  * Type EBS gp3
* OS: Ubuntu 22.04.3 LTS (Jammy Jellyfish)
* Kernel version: 6.2.0-1017-aws

**Comment:**

> We choose m5zn.2xlarge ec2 instance because it has consistent EBS performance unlike other similar (in terms of CPU and memory) instances which has baseline and bust EBS performance. 

Reviewer: ```suggestion
We choose m5zn.2xlarge ec2 instance because it has consistent EBS performance unlike other similar (in terms of CPU and memory) instances which have baseline and burst EBS performance. 
```

Author: Fixed. Thank you

Consistent EBS performance is important because we are using dedicated EBS disk for Longhorn storage. 

> If the EBS performance is inconsistent, we will see inconsistent test result overtime.

Reviewer: ```suggestion
If the EBS performance is inconsistent, we will see inconsistent test result over time.
```

Author: Fixed. Thank you

See more at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-optimized.html

### Network Spec
* Network bandwidth: 
  * Baseline bandwidth 10 Gbps
  * Burst bandwidth: Up to 25 Gbps in short time 
* Network latency: ~0.3ms RTT via ping command

### Disk Spec
* We are using dedicated disk for Longhorn volumes' replicas on the nodes. 
We select a typical EBS volume with average IOPs and bandwidth performance:
  * Single EC2 EBS gp3
  * 1TB 
  * IOPs set to 10000 
  * Throughput set to 360MiB/s 
  * Formatted as ext4 filesystem

**Comment:**
* We choose 10000 for EBS disk's IOPs simply because it is a middle value between minimum value 3000 and maximum value 16000 of the gp3 EBS disk
* We choose 360MiB/s for EBS disk's bandwidth because the m5zn.2xlarge EC2 instance has EBS bandwidth 396.25 MiB/s. 

>   If we choose a bigger value than 396.25 MiB/s for EBS disk's bandwidth, the ec2 instance would not be able to push EBS disk to that value.

Reviewer: NIT: Don't bother if it's too difficult or annoying, but links to where the reader can find this information might be useful.

Author: I decided to keep the info here since there isn't a single page which explain this behavior. It would be multiple links if we decided to put the links here

### Kubernetes Spec
* Kubernetes Version: v1.27.10+rke2r1
* CNI plugin: Calico
* Control plane nodes are separated from worker nodes

### Longhorn Config
* Longhorn version: v1.6.0
* Settings:
  * Using dedicated disk for Longhorn instead of root disk
  * The number of replicas per volume is 3
  * Storage Minimal Available Percentage setting: 10%
    * As we are using dedicated disk, we don't need big reserve storage as mentioned in best practice https://longhorn.io/docs/1.6.0/best-practices/#minimal-available-storage-and-over-provisioning
  * Storage Over Provisioning Percentage setting: 110%
    * We are planning to fill 15GB for each 20GB volume.
      If we schedule maximum amount, it would be 1100 GiB and actual usage will be (15/20)*1200 = 825GiB. This leaves 100GiB as 10% Storage Minimal Available Percentage setting plus some volumes' filesystem space overhead.

### Additional Components
* We deployed [Rancher monitoring](https://ranchermanager.docs.rancher.com/integrations-in-rancher/monitoring-and-alerting) which is a downstream version of [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack).

>   Note that this monitoring system generally consume certain amount of CPU, memory, and disk space on the node. We are deploying it with:

Reviewer: ```suggestion
  Note that this monitoring system generally consumes a certain amount of CPU, memory, and disk space on the node. We are deploying it with:
```

Author: Fixed. Thank you

  * CPU request: 750m
  * CPU limit: 1000m
  * Memory request: 750Mi
  * Memory limit: 3000Mi
  * Data retention size: 25GiB
* We deployed [Local Path Provisioner](https://github.com/rancher/local-path-provisioner) to test baseline storage performance when using  local storage in each node directly.

## Workload design

### Read and Write Tests Design

> We use [Kbench](https://github.com/yasker/kbench) which a tool to test Kubernetes storage.

Reviewer: ```suggestion
We use [Kbench](https://github.com/yasker/kbench), which is a tool to test Kubernetes storage.
```

Author: Fixed. Thank you

The idea of Kbench is that it deploys a pod with a Longhorn volume.

> Then it run various `fio` job (specified by user) to test multiple performance aspects of the volume (IOPs, bandwidth, latency).

Reviewer: ```suggestion
Then it runs various `fio` jobs (specified by user) to test multiple performance aspects of the volume (IOPs, bandwidth, latency).
```

Author: Fixed. Thank you

The pod runs the test repeatedly and exposes the result as Prometheus metrics.
We then collect and visualize the data using Grafana.

> Traditionally, kbench only deploy 1 single pod to test 1 volume. 

Reviewer: ```suggestion
Traditionally, kbench only deploys 1 single pod to test 1 volume. 
```

Author: Fixed. Thank you

However, in this report, we will gradually scale up the number of Kbench pods.
Since each pod stress-tests its Longhorn volume, as the number of Kbench pods go up,
we can simulate the situation in which the cluster has many pods doing IO aggressively. 
From there we can see the performance characteristic of Longhorn storage system as there are more and more IO intensive workloads.

We also perform the tests in which Kbench pods are rate-limited. 
From the rate-limited tests, we can further verify the result from the stress tests.

### Longhorn Control Tests Design

> The read and write performance tests above mostly care about Longhorn data plane. 

Reviewer: ```suggestion
The read and write performance tests above mostly concern the Longhorn data plane. 
```

Author: Fixed. Thank you

> In order to discover the scalability of Longhorn control plane, we also run tests that has non-IO workloads. 

Reviewer: ```suggestion
In order to discover the scalability of Longhorn control plane, we also run tests that have non-IO workloads. 
```

Author: Fixed. Thank you

These tests will help us to answer questions such as:
1. Maximum number of Longhorn volumes inside the cluster
1. Maximum number of Longhorn volumes attached to a node

> 1. Maximum number of Longhorn replicas on to a node

Reviewer: ```suggestion
1. Maximum number of Longhorn replicas on a node
```

Author: Fixed. Thank you

### Maximum Volume Actual Size Tests Design
From the experience and code analysis, the maximum Longhorn actual volume size is limited by the replica rebuilding time.
In this report, we will measure the replica rebuilding time as the size of Longhorn volume is getting bigger.
From there user can estimate the maximum volume size they can set given the information about how Longhorn it would take if Longhorn need to rebuild replicas for that volume.

### Load Balancing
In this report, we will distribute the load evenly across worker nodes:

1. Each worker nodes will have relative similar number of Kbench pods
2. Each worker nodes will have relative similar number of Longhorn replicas

This setup optimize efficiency of the cluster and Longhorn system. In practice, this balance is also what users usually strive for.

## Read Performance

### Random Read IOPs - Stress Tests

#### 1 control plane node + 3 worker nodes

We start with the cluster that has 1 control plane node + 3 worker nodes.

First, we do a comparison between a single RWO Longhorn PVC against a single Local Path Provisioner volume.

> We deploy 1 Kbench pod which attached a Local Path Provisioner PVC.

Reviewer: ```suggestion
We deploy 1 Kbench pod which attaches a Local Path Provisioner PVC.
```

Author: Fixed. Thank you

Then we delete the above Kbench pod and PVC and repeat the test with Longhorn PVC instead.

We use this yaml manifest for the Kbench workload:

> <details>

Reviewer: Cool use of `<details>`!

                - -c
                - '[ "$(ls -A /test-result)" ]'
            initialDelaySeconds: 30
            periodSeconds: 10
          command:
          - metric-exporter
          - -d
          - start
          env:
            - name: DATA_DIR
              value: /test-result
            - name: VOLUME_ACCESS_MODE
              value: rwo
            - name: TEST_MODE
              value: read-only
            - name: RATE_LIMIT_TYPE
              value: no-rate-limit
          ports:
            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: random-read-iops`: specify that we are running random read iops job. # TODO add the link to the job here

> * `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect.

Reviewer: We say this a lot throughout the document, but don't discuss how we know what size will avoid cache effect in the test design section. Can we improve this?

Author: FIxed. Thank you

* PVC size is 20G.

> Result: 
> * Local Path Provisioner: 10000
> * Longhorn: 22180

<img src="assets/medium-node-spec/read-performance/random-read-iops-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**: 

> * Because Longhorn has 3 replicas, it can read from 3 replicas concurrently thus may produce better read performance

Reviewer: ```suggestion
* Because Longhorn has 3 replicas, it can read from 3 replicas concurrently, and thus may produce better read performance.
```

Author: Fixed. Thank you

> * Total random IOPs can be achieved by all 3 Longhorn volumes is 31917

<img src="assets/medium-node-spec/read-performance/random-read-iops-1-3.png" alt="drawing" style="width:600px;"/>

**Comment**:
* Since each EBS volume on the host is provisioned with 10000 IOPs. Total IOPs of 3 EBS volumes is around 30000
* It looks like Longhorn system is able to reach the maximum IOPs capacity of the 3 host EBS volumes. 

Scaling workload from 3 to 6, then 6 to 9, then 9 to 12, then 12 to 15
> Result:
> * At 6 pods, the average random read IOPs per Longhorn volume is 5320. Total random IOPs is 31920
> * At 9 pods, the average random read IOPs per Longhorn volume is 3541. Total random IOPs is 31869
> * At 12 pods, the average random read IOPs per Longhorn volume is 2633. Total random IOPs is 31596
> * At 15 pods, the average random read IOPs per Longhorn volume is 2116. Total random IOPs is 31740

<img src="./assets/medium-node-spec/read-performance/random-read-iops-6-9-12-15.png" alt="drawing" style="width:600px;"/>

Note: the areas with the while background in the graph are transition periods. We can ignore the data in these areas. 

**Comment**:

> * From the scaling test so far, we can see that the total random read IOPs of all Longhorn volumes remain relative same around 31500 when the number of Kbench pods increase.

Reviewer: ```suggestion
* From the scaling test so far, we can see that the total random read IOPs of all Longhorn volumes remains relatively the same (around 31500) when the number of Kbench pods increase.
```

Author: Fixed. Thank you

  If we call the average random read IOPs each volume can achieve (x) and the number of volumes (y), they form a reciprocal function: x * y = 31500. 
  Users can use this information to make some prediction for this cluster:

>   * The upper bound limit that Longhorn system can achieve in this cluster is The value 31500 random read IOPs

Reviewer: ```suggestion
  * The upper bound limit that Longhorn system can achieve in this cluster is 31500 random read IOPs
```

Author: Fixed. Thank you

>   * If each of your workload pod is doing 1000 random IOPs in average, you can have estimatedly 31 pods

Reviewer: NIT: You haven't really introduced "you" before now. Better to talk about the pods or the user.

E.g.: "If each workload pod is doing 1000 random IOPs on average, there can be about 31 pods."

Author: Fixed. Thank you

> * When the user keeps scaling up number of pods eventually, this reciprocal relation (x * y = 31500) might no longer hold as the CPU contention and other factors kick in (i.e. x*y will be less and less)

Reviewer: ```suggestion
* When the user keeps scaling up number of pods, eventually this reciprocal relation (x * y = 31500) might no longer hold as the CPU contention and other factors kick in (i.e. x*y will be less and less)
```

Author: Fixed. Thank you

          ports:
            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: sequential-read-iops`: specify that we are running sequential read iops job. # TODO add the link to the job here
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect.
* PVC size is 20G.

> Result:
> * Local Path Provisioner: 10000
> * Longhorn: 39772

<img src="./assets/medium-node-spec/read-performance/sequential-read-iops-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comments:**
* We observe Longhorn volume achieves much bigger sequential read IOPs compares to the local path volume. 
  This can be explained by 2 factors:

>   * Longhorn volume has 3 replicas so it can read concurrently from multiple replicas (each replica is on a different EBS disk) while the local path volume is reading from single EBS disk.

Reviewer: ```suggestion
  * Each Longhorn volume has 3 replicas so it can read concurrently from multiple replicas (each replica is on a different EBS disk) while the local path volume is reading from single EBS disk.
```

Author: Fixed. Thank you

>   * When running Kbench with Longhorn volume, there is IO merging happens at both kernel layer and EBS layer. 

Reviewer: ```suggestion
  * When running Kbench with Longhorn volume, there is IO merging at both the kernel layer and the EBS layer. 
```

Author: Fixed. Thank you

> Result:
> * At 6 pods, the average sequential read IOPs per Longhorn volume is 10132. Total sequential read IOPs is 60792
> * At 9 pods, the average sequential read IOPs per Longhorn volume is 6875. Total sequential read IOPs is 61875
> * At 12 pods, the average sequential read IOPs per Longhorn volume is 5210. Total sequential read IOPs is 62520
> * At 15 pods, the average sequential read IOPs per Longhorn volume is 4159. Total sequential read IOPs is 62385

<img src="./assets/medium-node-spec/read-performance/sequential-read-iops-6-9-12-15.png" alt="drawing" style="width:600px;"/>

Note: the areas with the while background in the graph are transition periods. We can ignore the data in these areas.

**Comment**:
* From the scaling test so far, we can see that the total sequential read IOPs of all Longhorn volumes remain relative same around 61500 when the number of Kbench pods increase.
  If we call the average sequential read IOPs each volume can achieve (x) and the number of volumes (y), they form a reciprocal function: x * y = 61500.
  Users can use this information to make some prediction for this cluster:

>     * The upper bound limit that Longhorn system can achieve in this cluster is The value 61500 sequential read IOPs

Reviewer: ```suggestion
    * The upper bound limit that Longhorn system can achieve in this cluster is 61500 sequential read IOPs
```

Author: Fixed. Thank you

          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

> Result:
> * The sequential read IOPs per Longhorn volume is 1190
> * The total sequential read IOPs can be achieved by all 51 Longhorn volumes is 60690

<img src="./assets/medium-node-spec/read-performance/sequential-read-iops-rate-limited-1205.png" alt="drawing" style="width:600px;"/>

**Comments:**
* As discussed in the [Sequential Read IOPs - Stress Tests](#sequential-read-iops---stress-tests) section above, we come up with the conclusion that
  `If we call the sequential read IOPs each volume can achieve (x) and the number of volumes (y), they form a reciprocal function: x * y = 61500.`
* The test result here further confirm this conclusion as x * y = 1190*51 = 60690 (slightly lower than 61500 but it is insignificant difference)

> ### Random Read Bandwidth - Stress Tests

Reviewer: Stopped here for now. Will continue later.

          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:
* `MODE: random-read-bandwidth`: specify that we are running random read bandwidth job. # TODO add the link to the job here
* `SIZE: 15G`: the `fio` test size is 15G, this will avoid cache effect.
* PVC size is 20G.

> Result:
> * Local Path Provisioner: 362 MiB/s
> * Longhorn: 874 MiB/s

<img src="./assets/medium-node-spec/read-performance/random-read-bandwidth-local-path-vs-longhorn.png" alt="drawing" style="width:600px;"/>

**Comment**:
* Because Longhorn has 3 replicas, it can read from 3 replicas concurrently thus may produce better read performance

Next, we use Kbench with Longhorn PVCs and scale up the number of Kbench pods to see how random read bandwidth is effected when there are more and more IO intensive pods.

Scaling workload from 1 to 3 pods.

> Result:
> * Each Kbench pod is able to achieve 386 MiB/s random read bandwidth on its Longhorn volume 

> > * Total random read bandwidth can be achieved by all 3 Longhorn volumes is 1158

Reviewer: ```suggestion
> * Total sequential read bandwidth can be achieved by all 3 Longhorn volumes is 1158
```

Author: I think it should be `random`

<img src="./assets/medium-node-spec/read-performance/random-read-bandwidth-1-3.png" alt="drawing" style="width:600px;"/>

**Comment:**
* Since each EBS volume on the host is provisioned with 360 MiB/s. Total IOPs of 3 EBS volumes is around 1080 MiB/s
It looks like Longhorn system is able to reach the maximum IOPs capacity of the 3 host EBS volumes. 
Longhorn is able to achieve slightly above 1080 MiB/s maybe due to small number of IO mergings or EBS actual bandwidth might be a bit higher the specified limit. 

Scaling workload from 3 to 6, then 6 to 9, then 9 to 12, then 12 to 15

> Result:

> > * At 6 pods, the average random read bandwidth per Longhorn volume is 196 MiB/s. Total random bandwidth is 1176 MiB/s

Reviewer: ```suggestion
> * At 6 pods, the average sequential read bandwidth per Longhorn volume is 196 MiB/s. Total random bandwidth is 1176 MiB/s
```

And below.

Author: I think it should be `random`

        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

> Result:
> * The random read bandwidth per Longhorn volume is 22.3 MiB/s
> * The total random read bandwidth can be achieved by all 51 Longhorn volumes is 1137 MiB/s

<img src="./assets/medium-node-spec/read-performance/random-read-bandwidth-rate-limited.png" alt="drawing" style="width:600px;"/>

**Comments:**
* As discussed in the [Random Read Bandwidth - Stress Tests](#random-read-bandwidth---stress-tests) section above, we come up with the conclusion that
  `If we call the random read bandwidth each volume can achieve (x) and the number of volumes (y), they form a reciprocal function: x * y = 1160.`
* The test result here further confirm this conclusion as x * y = 22.3*51 = 1137 (slightly lower than 1160 as the CPU contention and other factors kick in)

### Sequential Read Bandwidth - Stress Tests
#### 1 control plane node + 3 worker nodes
We start with the cluster that has 1 control plane node + 3 worker nodes.

First, we do a comparison between a single RWO Longhorn PVC against a single Local Path Provisioner volume.

> We deploy 1 Kbench pod which attached a Local Path Provisioner PVC.

Reviewer: ```suggestion
We deploy 1 Kbench pod which attaches a Local Path Provisioner PVC.
```

Author: Fixed. Thank you

          readinessProbe:
            exec:
              command:
                - sh
                - -c
                - '[ "$(ls -A /test-result)" ]'
            initialDelaySeconds: 30
            periodSeconds: 10
          command:
            - metric-exporter
            - -d
            - start
          env:
            - name: DATA_DIR
              value: /test-result
            - name: VOLUME_ACCESS_MODE
              value: rwo
            - name: TEST_MODE
              value: read-only
            - name: RATE_LIMIT_TYPE
              value: no-rate-limit
          ports:
            - containerPort: 8080
              name: metrics
          volumeMounts:
            - name: vol
              mountPath: /volume/
            - name: shared-data
              mountPath: /test-result
      volumes:
        - name: shared-data
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: vol
      spec:
        accessModes: ["ReadWriteOnce"]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 20Gi
```
</details>

Some of the important Kbench parameters we would like to call out are:

> * `MODE: sequential-read-bandwidth`: specify that we are running sequential read bandwidth job. # TODO add the link to the job here

Reviewer: ```suggestion
* `MODE: sequential-read-bandwidth`: specifies that we are running a sequential read bandwidth job. # TODO add the link to the job here
```

Can we complete the TODO?

Author: Thank you. Fixed

# Reference Setup, Performance, Scalability, and Sizing Guidelines: Public Cloud - Medium Node Spec

## Cluster Spec:
Node spec:
* EC2 instance type: ec2 m5.2xlarge
* 8vCPUs, 32GB RAM
* Root disk
  * Size 50GB 
  * Type EBS gp3
* OS: Ubuntu 22.04 LTS
* Kernel version: TODO

Network:
* Network bandwidth: up to 10 Gbits/sec
* Network latency: ~0.3ms RTT via ping command

Disk spec:
* We are using dedicated disk for Longhorn volumes' replicas on the nodes:

>   * 1TB block storage on each node.

Reviewer: Nit: Make it super clear that we are using a SINGLE dedicated disk (if that's what we are doing). I think it reads that way, but better to be clear.

Author: Good point! Thank you! Updated

>   * EC2 EBS gp3 with IOPS set to 6000 and throughput set to 250MiB/s

Reviewer: Question: In some of your preliminary 3 node results, you hit a cap of ~20000 IOPs (6000 x 3 + some). Is there a chance that the IOPs limit is a bottleneck?

Author: Good idea! That number was from digital ocean cluster but I will keep this in mind

Author: It is indeed related. I made comments in the report. Thank you
